Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs

```diff
@@ -49,10 +49,48 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
             return;
 
         if (GetBallisticShots(component) >= component.Capacity)
+        {
+            Popup(Loc.GetString("gun-ballistic-transfer-target-full", ("entity", args.Target)), uid, args.User);
             return;
+        }
+
+        if (EntityManager.HasComponent<SpeedLoaderComponent>(args.Used))
+        {
+            var emptySlots = component.Capacity - component.UnspawnedCount - component.Entities.Count; // Number of empty slots in the shotgun
+
+            var ammo = new List<(EntityUid? Entity, IShootable Shootable)>();
+            var ev = new TakeAmmoEvent(emptySlots, ammo, Transform(uid).Coordinates, args.User);
+            RaiseLocalEvent(args.Used, ev);
+
+            // empty speedloader
+            if (ev.Ammo.Count == 0)
+            {
+                Popup(Loc.GetString("gun-speedloader-empty"), uid, args.User);
+                return;
+            }
+
+            foreach (var (ent, _) in ammo)
+            {
+                if (ent == null)
+                    continue;
+
+                component.Entities.Add(ent.Value);
+                Containers.Insert(ent.Value, component.Container);
+
+                if (IsClientSide(ent.Value))
+                    Del(ent.Value);
+
+                if (ev.Ammo.Count == 0)
+                    break;
```

Comment:
The event one is just a ref to the local variable, is this even possible here? It's being iterated in the loop.

Reply:
It seems to work, is there another way to do this that's preferred or is that fine?

https://github.com/user-attachments/assets/571c6bb6-66fe-45de-a434-0fe8c3db1f4f